base/task.jl

```diff
@@ -479,14 +485,43 @@ macro async(expr)
         let $(letargs...)
             local task = Task($thunk)
             if $(Expr(:islocal, var))
-                put!($var, task)
+                put!($var, $(wrap(:task)))
             end
             schedule(task)
             task
         end
     end
 end
 
+# task wrapper that doesn't create exceptions wrapped in TaskFailedException
+struct UnwrapTaskFailedException
+    task::Task
+end
+
+# common code for wait&fetch for UnwrapTaskFailedException
+function unwrap_task_failed(f::Function, t::UnwrapTaskFailedException)
+    try
+        f(t.task)
+    catch ex
+        if ex isa TaskFailedException
+            throw(ex.task.exception)
+        else
+            rethrow()
+        end
+    end
+end
+
+# the unwrapping for above task wrapper (gets triggered in sync_end())
+wait(t::UnwrapTaskFailedException) = unwrap_task_failed(wait, t)
+
+# same for fetching the tasks, for convenience
+fetch(t::UnwrapTaskFailedException) = unwrap_task_failed(fetch, t)
+
+# macro for running async code that doesn't throw wrapped exceptions
+macro async_nowrap(expr)
```

Comment:
Just bike shedding: what do you think about `async_unwrap`?

Reply:
yes that's actually better name given the thing actually _unwraps_ the exception...

(perhaps, should we go all the way to `@async_unwrap_exceptions` so that the name is "full"? )

Reply:
All the way would be `async_unwrap_taskfailedexception`, I guess. That's too verbose for my taste, I think `async_unwrap` unwrap is fine. It's not going to be part of the public API, is it?

Reply:
ok, done.

> It's not going to be part of the public API, is it?

I guess not. I additionally `import`ed it into Distributed just as it was with `@sync_add`, for consistency.

Reply:
( re public API, some exposure could help for https://github.com/JuliaLang/julia/issues/38931 but that's for a different PR I'd say. )